### PR TITLE
Add an expand provider, currently expands only env variables

### PR DESCRIPTION
**Pull request by bogdandrutu**
Removes the logic from the "configunmarshaler" which should not have to deal with preparing the config.

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

### CODE REVIEW

File: service/parserprovider/expand.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+	"go.opentelemetry.io/collector/config"
+)
+
+type expandMapProvider struct {
+	base MapProvider
+}
+
+// NewExpandMapProvider returns a MapProvider, that expands all environment variables for a
+// config.Map provided by the given MapProvider.
+func NewExpandMapProvider(base MapProvider) MapProvider {
```
Comment (pjanotti) [inline service/parserprovider/expand.go:30]: Suggestion `NewEnvVarExpandMapProvider`, it is a big name but it makes clear what it expands.

File: service/parserprovider/expand.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+			nslice = append(nslice, expandStringValues(vint))
+		}
+		return nslice
+	case map[string]interface{}:
+		nmap := make(map[interface{}]interface{}, len(v))
+		for k, vint := range v {
+			nmap[k] = expandStringValues(vint)
+		}
+		return nmap
+	case map[interface{}]interface{}:
```
Comment (pjanotti) [inline service/parserprovider/expand.go:69]: Are you sure that we still hit this case? IIRC after we switched to koanf this case may not be hit.

**Review by pjanotti — APPROVED** [review]
LGTM, just a minor name suggestion and one Q.

File: service/parserprovider/expand.go (hunk 1 above)
Comment (bogdandrutu) [reply service/parserprovider/expand.go:30]: I think this should be merged with the config source manager later. That's why I kept a different name.

File: service/parserprovider/expand.go (hunk 2 above)
Comment (bogdandrutu) [reply service/parserprovider/expand.go:69]: Not sure, but trying to limit the changes in one PR

File: service/parserprovider/testdata/expand-escaped-env.yaml [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  # $$$$ -> two escaped $
+  recv.3: "$$$$MAP_VALUE_3"
+  # escaped $ in the middle
+  recv.4: "some$${MAP_VALUE_4}text"
+  # two escaped $
+  recv.5: "$${ONE}$${TWO}"
+  # trailing escaped $
+  recv.6: "text$$"
+  # escaped $ alone
+  recv.7: "$$"
```
Comment (codeboten) [inline service/parserprovider/testdata/expand-escaped-env.yaml:15]: What happens in the case of a single `$`?

File: service/parserprovider/expand_test.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+			"recv.5": "${ONE}${TWO}",
+			// trailing escaped $
+			"recv.6": "text$",
+			// escaped $ alone
+			"recv.7": "$",
+		}}
+	assert.Equal(t, expectedMap, cfgMap.ToStringMap())
+}
+
+// TODO: When FileProvider supports passing the file name remove this.
```
Comment (codeboten) [inline service/parserprovider/expand_test.go:102]: Please open an issue if one doesnt already exist.

File: service/parserprovider/testdata/expand-escaped-env.yaml (hunk 3 above)
Comment (bogdandrutu) [reply service/parserprovider/testdata/expand-escaped-env.yaml:15]: That is the normal escaped value

Comment (bogdandrutu) [reply service/parserprovider/testdata/expand-escaped-env.yaml:15]: It is tested in all other tests.

File: service/parserprovider/expand_test.go (hunk 4 above)
Comment (bogdandrutu) [reply service/parserprovider/expand_test.go:102]: Depending on https://github.com/open-telemetry/opentelemetry-collector/pull/4077 which one gets merge first the second will fix this.

**Review by codeboten — APPROVED** [review]

**Pull request merged**